### If문, try/catch문은 무엇이 다른가?


### 0. 들어가면서

* if 조건문을 쓰는 이유는 무엇인가?
* try/catch문을 쓰는 이유는 무엇인가?
  * 나는 Java뿐만 아니라 다른 언어에서 지금까지 코드를 짜면서 거의 대부분의 조건문을 if문으로 해결했다. (try/catch문이 더 어울리는 상황에서도...)물론 다른 사람의 코드를 참고하거나, 책을 보거나, 학교 수업을 따라갈 때는 간혹 try/catch문을 이용하기도 했다. 하지만 정작 try/catch문을 왜 꼭 써야 하는가에 대해서는 명확히 이유를 제시하지 못했다. Exception을 처리하기 위해서 try/catch문을 쓴다(?) 정도의 얄팍한 수준. 이번 글에서는 if문과 try/catch문의 차이에 대해서 알아보자.



### 1. 예외 처리

* Try/Catch문을 사용하는 가장 큰 목적은 **예외 처리**를 하는 데 있다. 예외 처리(Exception Handling)이란 두 가지 방식으로 수행할 수 있는데, 첫번째가 전통적인 방식인 조건문을 이용하는 것이고, 두번째 방법이 try-catch문을 이용하는 것이다.

* **그렇다면 if문과 try-catch문은 무엇이 다른가?**

  * try-catch문에서는 예외가 발생했을 때 그 즉시 블록이 종료되고 catch 블록으로 제어가 이동한다. 하지만 if-else로 에러를 처리할 경우에는 에러가 발생한 객체에 대하여 수명이 유지되기 때문에 에러를 처리하는 동안에도 에러를 발생한 객체를 참조하는 코드가 정상적으로 컴파일되어 위험한 상황이 발생할 수 있는 것이다. 

  * 또한, try/catch문은 예외가 발생했을 때 "문제가 발생할 당시의 파라미터", "해당 예외에 대한 설명" 등을 전달해줌으로써 정보 전달의 기능도 할 수 있다. 쉽게 말해서 어디서 프로그램이 터진 것인지 쉽게 파악할 수 있는 것이다.

  * 그리고 무엇보다, 지금 우리는 **"에러 처리"가 아닌 "예외 처리"**를 하고 있음에 주목해야 한다. 에러는 프로그램 코드에 의해서 수습될 수 없는 오류가 나는 것이고, 예외는 프로그램 코드에 의해 수습될 수 있는 다소 미약한 오류라고 생각할 수 있다. 예외는 예외 처리를 통해 프로그램을 종료하지 않고 정상적으로 작동되게 만들어 줄 수 있다.

    

### 2. try/catch문의 기본 구조

```java
try{
    예외가 생길 가능성이 있는 코드 작성
}catch(예외 발생 클래스명e){
    예외처리 코드
}finally...
```

* 구체적인 소스 코드를 한 번 보자.

 ```java
 public class Math{
     public static void main(String[] args){
         int num1, num2;
         Scanner sc = new Scanner(System.in); # import 생략
         num1 = sc.nextInt();
         num2 = sc.nextInt();

         try{
             System.out.println(num1/num2);
         }catch (Exception e){
             System.out.println("0으로 나눌 수 없습니다.");

         }
     }
 }
 ```
* 위 코드에서 사용자가 num2에 0을 대입했을 때는 "0으로 나눌 수 없습니다."라는 메시지가 출력된다. 하지만 try-catch문이 없을 경우에는 ArithmeticException이라는 이름의 에러가 발생하게 된다. 이런식으로 직접적으로 어떤 error가 발생하는지 Exception의 클래스명을 알고 있다면 각각의 에러에 대하여 예외처리를 할 수 있게 되는 것이다.

* 그렇게 되면 여러 에러가 발생할 가능성을 가지고 각각의 에러에 대한 예외처리 구문을 작성해 줄 수 있게 되고, 코드를 작성한 사람이나 다른 사람이 코드를 보았을 때 한번에 파악하기도 쉬워질 것이다.

 ```java
 public class Math{
     public static void main(String[] args){
         int num1, num2;
         Scanner sc = new Scanner(System.in); # import 생략
         num1 = sc.nextInt();
         num2 = sc.nextInt();

         try{
             System.out.println(num1/num2);
         }catch (ArithmeticException e){ # <- 이런 식으로 
             System.out.println("0으로 나눌 수 없습니다.");
         }
     }
 }
 ```

### 3. try/catch문의 디테일 - 예외에 의미를 제공할 것
* 예외를 던질 때는 전후 상황을 충분히 덧붙여야 한다. 그러면 오류가 발생한 원인과 위치를 찾기가 쉬워진다. 자바는 모든 예외에 호출 스택을 제공한다. 하지만 실패한 코드의 의도를 파악하려면 호출 스택만으로는 부족하다. 오류 메시지에 정보를 담아 예외와 함께 던져야 한다. 위의 1번에서 이야기한 것처럼 실패한 연산 이름과 실패 유형도 언급하는 것이 좋다. 


### 4. try/catch문의 디테일 - 호출자를 고려해 예외 클래스를 정의할 것
* 오류를 분류하는 방법은 수없이 많다. 오류가 발생한 위치로 분류가 가능하다. 예를 들어, 오류가 발생한 컴포넌트로 분류한다. 아니면 유형으로도 분류가 가느앟다. 예를 들어 디바이스 실패, 네트워크 실패, 프로그래밍 오류 등으로 분류한다. 하지만 애플리케이션에서 오류를 정의할 때 프로그래머에게 가장 중요한 관심사는 오류를 잡아내는 방법이 되어야 한다.

* 아래 코드는 예외를 형편없이 분류한 예시다.
```java
ACMPort port = new ACMPort(12);

try{
  port.open();
} catch (DeviceResponseException e){
    reportPortError(e);
    logger.log("Device response exception", e);
} catch (ATM1212UnlockedException e){
    reportPortError(e);
    logger.log("Unlock exception", e);
} catch (GMXError e){
    reportPortError(e);
    logger.log("Device response exception", e);
} finally{
    ...
}
```
* 위 경우는 예외에 대응하는 방식이 예외 유형과 무관하게 거의 똑같다. 그래서 코드를 간결하게 고치기가 아주 쉽다. 호출하는 라이브러리 API를 감싸면서 예외 유형 하나를 반환하는 것이다.
```java
LocalPort port = new LocalPort(12);
try{
  port.open();
} catch (PortDeviceFailure e) {
  reportError(e);
  logger.log(e.getMessage(), e);
} finally{
  ...
}
```
* 여기서 LocalPort 클래스는 단순히 ACMEPort 클래스가 던지는 예외를 잡아 변환하는 Wrapper 클래스일 뿐이다.
```java
public class LocalPort{
  private ACMEPort innerport;
  
  public LocalPort(int portNumber){
    innerPort = new ACMEPort(portNumber);
  }
  
  public void open(){
    try{
      innerport.open();
    } catch (DeviceResponseEexception e){
      throw new PortDeviceFailure(e);
    } catch (ATM1212UnlockedException e){
      throw new PortDeviceFailure(e);
    } catch (GMXError e){
      throw new PortDeviceFailure(e);
    }
  }
  ...
}
```
* Wrapper 클래스를 예외 처리에 사용하면 코드가 간결해지고 정확히 어떤 예외를 처리하려고 하는지 중복없이 깔끔하게 전달할 수 있다. 또한 외부 API를 감싸면 외부 라이브러리와 프로그램 사이에서 의존성이 크게 줄어든다. 나중에 다른 라이브러리로 갈아타도 유지 보수가 쉬워지는 것이다.


### 5. 예외 처리의 중요성
* 프로그램에서 예외는 발생할 수밖에 없고, 특히 예외로 인해 서버 프로그램이 작동을 멈춘다면 문제가 커진다. 적어도 발생한 예외가 어떤 예외이며, 이것이 무시해도 될 수준인지 아니면 반드시 처리를 해야하는 것인지 그것도 아니면 시스템을 종료시켜야 하는지에 대해 생각해봐야 한다.
* 프로그램을 설계하고 구축하는 첫 단계에서 많은 예외 상황을 고려해보고 이를 어떻게 전달하고 처리할지 고민해보는 습관을 들여야 하겠다.
